File: fe_camphub/src/libs/core/mapping/transaction.mapper.ts
import { createMap, createMapper } from "@automapper/core";
import { pojos, PojosMetadataMap } from "@automapper/pojos";
import { Transaction, TransactionBooking } from "../types";
import { TransactionResponse, TransactionDetailResponse } from "../dto/response";

// ------------------------
// Mapper
// ------------------------
export const transactionMapper = createMapper({
    strategyInitializer: pojos(),
});

// ------------------------
// Metadata
// ------------------------
PojosMetadataMap.create<Transaction>("Transaction", {
    id: String,
    fromAccountId: String,
    toAccountId: String,
    amount: Number,
    type: String,
    status: String,
    createdAt: String,
    senderName: String,
    receiverName: String,
});
PojosMetadataMap.create<TransactionResponse>("TransactionResponse", {
    id: String,
    fromAccountId: String,
    toAccountId: String,
    amount: Number,
    type: String,
    status: String,
    createdAt: String,
    senderName: String,
    receiverName: String,
});
PojosMetadataMap.create<TransactionBooking>("TransactionBooking", {
    transactionId: String,
    amount: Number,
    type: String,
    status: String,
    createdAt: String,

    bookingId: String,
    itemId: String,
    itemName: String,
    lesseeId: String,
    lessorId: String,

    lesseeName: String,
    lessorName: String,
});
PojosMetadataMap.create<TransactionDetailResponse>("TransactionDetailResponse", {
    transactionId: String,
    amount: Number,
    type: String,
    status: String,
    createdAt: String,

    bookingId: String,
    itemId: String,
    itemName: String,
    lesseeId: String,
    lessorId: String,

    lesseeName: String,
    lessorName: String,
});

// ------------------------
// Mapping: DTO response → Model
// ------------------------
createMap<TransactionResponse, Transaction>(
    transactionMapper,
    "TransactionResponse",
    "Transaction"
);

createMap<TransactionDetailResponse, TransactionBooking>(
    transactionMapper,
    "TransactionDetailResponse",
    "TransactionBooking"
);

// ------------------------
// Export API
// ------------------------
export const transactionMap = {
    fromResponse: (response: TransactionResponse): Transaction =>
        transactionMapper.map<TransactionResponse, Transaction>(
            response,
            "TransactionResponse",
            "Transaction"
        ),
    fromDetailResponse: (response: TransactionDetailResponse): TransactionBooking =>
        transactionMapper.map<TransactionDetailResponse, TransactionBooking>(
            response,
            "TransactionDetailResponse",
            "TransactionBooking"
        ),
};

File: be_camphub/src/main/java/org/camphub/be_camphub/mapper/TransactionMapper.java
package org.camphub.be_camphub.mapper;

import org.camphub.be_camphub.dto.response.transaction.TransactionDetailResponse;
import org.camphub.be_camphub.dto.response.transaction.TransactionResponse;
import org.camphub.be_camphub.entity.Booking;
import org.camphub.be_camphub.entity.Item;
import org.camphub.be_camphub.entity.Transaction;
import org.camphub.be_camphub.entity.TransactionBooking;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring")
public interface TransactionMapper {
    @Mapping(target = "senderName", ignore = true) // sẽ set trong service
    @Mapping(target = "receiverName", ignore = true) // sẽ set trong service
    TransactionResponse entityToResponse(Transaction transaction);

    @Mapping(target = "transactionId", source = "transaction.id")
    @Mapping(target = "amount", source = "transaction.amount")
    @Mapping(target = "type", source = "transaction.type")
    @Mapping(target = "status", source = "transaction.status")
    @Mapping(target = "createdAt", source = "transaction.createdAt")
    @Mapping(target = "bookingId", source = "booking.id")
    @Mapping(target = "itemId", source = "item.id")
    @Mapping(target = "itemName", source = "item.name")
    @Mapping(target = "lesseeId", source = "booking.lesseeId")
    @Mapping(target = "lessorId", source = "booking.lessorId")
    @Mapping(target = "lesseeName", ignore = true) // sẽ set trong service
    @Mapping(target = "lessorName", ignore = true) // sẽ set trong service
    TransactionDetailResponse toTransactionDetailResponse(Transaction transaction, TransactionBooking transactionBooking, Booking booking, Item item);
}
